AmplifyPlugins/DataStore/AWSDataStoreCategoryPluginIntegrationTests/DataStoreEndToEndTests.swift

```diff
@@ -366,6 +367,175 @@ class DataStoreEndToEndTests: SyncEngineIntegrationTestBase {
 
     }
 
+    /// Ensure DataStore.stop followed by DataStore.start is successful
+    ///
+    /// - Given:  DataStore has just configured, but not started
+    /// - When:
+    ///    - DataStore.stop
+    ///    - Followed by DataStore.start in the completion of the stop
+    /// - Then:
+    ///    - Datastore should be started successfully
+    ///
+    func testConfigureAmplifyThenStopStart() throws {
+        setUp(withModels: TestModelRegistration())
+        try startAmplify()
+        let stopStartSuccess = expectation(description: "stop then start successful")
+        Amplify.DataStore.stop { result in
+            switch result {
+            case .success:
+                Amplify.DataStore.start { result in
+                    switch result {
+                    case .success:
+                        stopStartSuccess.fulfill()
+                    case .failure(let error):
+                        XCTFail("\(error)")
+                    }
+                }
+            case .failure(let error):
+                XCTFail("\(error)")
+            }
+        }
+        wait(for: [stopStartSuccess], timeout: networkTimeout)
+
+    }
+
+    /// Ensure DataStore.clear followed by DataStore.start is successful
+    ///
+    /// - Given:  DataStore has just configured, but not started
+    /// - When:
+    ///    - DataStore.clear
+    ///    - Followed by DataStore.start in the completion of the stop
+    /// - Then:
+    ///    - Datastore should be started successfully
+    ///
+    func testConfigureAmplifyThenClearStart() throws {
+        setUp(withModels: TestModelRegistration())
+        try startAmplify()
+        let clearStartSuccess = expectation(description: "clear then start successful")
+        Amplify.DataStore.clear { result in
+            switch result {
+            case .success:
+                Amplify.DataStore.start { result in
+                    switch result {
+                    case .success:
+                        clearStartSuccess.fulfill()
+                    case .failure(let error):
+                        XCTFail("\(error)")
+                    }
+                }
+            case .failure(let error):
+                XCTFail("\(error)")
+            }
+        }
+        wait(for: [clearStartSuccess], timeout: networkTimeout)
+
+    }
+
+    /// Ensure DataStore.stop followed by DataStore.start is successful on
+    ///  a DispatchQueue
+    ///
+    /// - Given:  DataStore has just configured, but not started
+    /// - When:
+    ///    - DataStore.stop
+    ///    - Followed by DataStore.start in the completion of the stop
+    /// - Then:
+    ///    - Datastore should be started successfully
+    ///
+    func testConfigureAmplifyThenStopStartonDispatch() throws {
```

Comment:
what are we testing here by running the code on the background thread? should we just remove this test?

Reply:
the workaround for the customer without this PR where the DispatchQueue.async is performed before the second call, not before both calls, like this:
```
DataStore.stop {   
  DispatchQueue.async {
      DataStore.start { 
       }
   }
}
```

We can provide this info to the customer

Reply:
I feel this test is to make sure that running consecutive start-stop on a single thread (DispatchQueue) 🧵 is still successful. 
I can remove this if we don't require!

Reply:
i don't think this test is required because whether we decide to run it on a new queue or run it in current executing thread of the test, it's still just a thread we're running on. I test that may be useful to have is to have the workaround as a test (see second comment of this comment thread) to show that the implementation changes in this PR does not cause the workaround to no longer work